internal/conversions/v1beta2.go

```diff
@@ -38,343 +33,34 @@ func ownerFromInternalV1beta2(in *internalversion.Owner, out *v1beta2.Owner, _ c
-		Retries:  h.Retries,
-	}
-}
-
-func fromComposePlacement(p composetypes.Placement) v1beta2.Placement {
-	return v1beta2.Placement{
-		Constraints: fromComposeConstraints(p.Constraints),
-	}
-}
-
-var constraintEquals = regexp.MustCompile(`([\w\.]*)\W*(==|!=)\W*([\w\.]*)`)
-
-const (
-	swarmOs          = "node.platform.os"
-	swarmArch        = "node.platform.arch"
-	swarmHostname    = "node.hostname"
-	swarmLabelPrefix = "node.labels."
-)
-
-func fromComposeConstraints(s []string) *v1beta2.Constraints {
-	if len(s) == 0 {
-		return nil
-	}
-	constraints := &v1beta2.Constraints{}
-	for _, constraint := range s {
-		matches := constraintEquals.FindStringSubmatch(constraint)
-		if len(matches) == 4 {
-			key := matches[1]
-			operator := matches[2]
-			value := matches[3]
-			constraint := &v1beta2.Constraint{
-				Operator: operator,
-				Value:    value,
-			}
-			switch {
-			case key == swarmOs:
-				constraints.OperatingSystem = constraint
-			case key == swarmArch:
-				constraints.Architecture = constraint
-			case key == swarmHostname:
-				constraints.Hostname = constraint
-			case strings.HasPrefix(key, swarmLabelPrefix):
-				if constraints.MatchLabels == nil {
-					constraints.MatchLabels = map[string]v1beta2.Constraint{}
-				}
-				constraints.MatchLabels[strings.TrimPrefix(key, swarmLabelPrefix)] = *constraint
-			}
-		}
-	}
-	return constraints
-}
-
-func fromComposeResources(r composetypes.Resources) v1beta2.Resources {
-	return v1beta2.Resources{
-		Limits:       fromComposeResourcesResource(r.Limits),
-		Reservations: fromComposeResourcesResource(r.Reservations),
-	}
-}
-
-func fromComposeResourcesResource(r *composetypes.Resource) *v1beta2.Resource {
-	if r == nil {
-		return nil
-	}
-	return &v1beta2.Resource{
-		MemoryBytes: int64(r.MemoryBytes),
-		NanoCPUs:    r.NanoCPUs,
-	}
-}
-
-func fromComposeUpdateConfig(u *composetypes.UpdateConfig) *v1beta2.UpdateConfig {
-	if u == nil {
-		return nil
-	}
-	return &v1beta2.UpdateConfig{
-		Parallelism: u.Parallelism,
-	}
-}
-
-func fromComposeRestartPolicy(r *composetypes.RestartPolicy) *v1beta2.RestartPolicy {
-	if r == nil {
-		return nil
-	}
-	return &v1beta2.RestartPolicy{
-		Condition: r.Condition,
-	}
-}
-
-func fromComposeServiceVolumeConfig(vs []composetypes.ServiceVolumeConfig) []v1beta2.ServiceVolumeConfig {
-	if vs == nil {
-		return nil
-	}
-	volumes := []v1beta2.ServiceVolumeConfig{}
-	for _, v := range vs {
-		volumes = append(volumes, v1beta2.ServiceVolumeConfig{
-			Type:     v.Type,
-			Source:   v.Source,
-			Target:   v.Target,
-			ReadOnly: v.ReadOnly,
-		})
-	}
-	return volumes
-}
-
-// StackFromInternalV1beta2 converts an internal stack to v1beta2 flavor
-func StackFromInternalV1beta2(in *internalversion.Stack) (*v1beta2.Stack, error) {
-	if in == nil {
-		return nil, nil
-	}
-	out := new(v1beta2.Stack)
-	err := stackFromInternalV1beta2(in, out, nil)
-	return out, err
+	return v1alpha3.Convert_v1alpha3_Stack_To_v1beta2_Stack(&alpha3, out, s)
```

Comment:
I think this is not very future proof, as you don't really know yet how alpha3 will evolve... Now an alpha3 is a beta2, but it won't be valid soon I guess...

Reply:
v1alpha3 will be a superset of beta2, and it will be the "most complete" format. So using it as a pivot version seemed to be a good idea